### PR TITLE
Cm/try playwright

### PR DESCRIPTION
**Pull request by cmishra**
## Goals
 - Does controlling chrome with selenium lead to better reliability?
 - Is playwright more reliable than selenium? Is it faster? 
 - Is playwright better for developer productivity?

Previous PR: https://github.com/instabase/task-tracker/pull/11 

## Results
**tldr**:  playwrite + firefox + docker has a very good success rate (80% - 90%) with a ~3x improvement in latency vs selenium (18 seconds for passing, 35 seconds for failing runs). developer productivity is also dramatically better with it with better debugging/logging, concise syntax, and no need for explicit waits. 

 - Background review: Firefox+selenium had a 60% success rate (with/without docker), ~45 seconds latency when successful ~75 seconds when failed (for the median). 
 - Playwright was much better from a developer productivity perspective. Better error logs, much more succinct syntax, no need for a bunch of explicit wait statements (when you will also be interacting with that element you're waiting for). 
 - Chrome+selenium doesn't lead to more reliability (~65% failure rate) but is much faster than the firefox +selenium:
```bash
Chetans-MacBook-Pro:browser-automation chetanmishra$ python scripts/generate_summary_stats.py --path selenium_results/benchmark-n-50-no-docker-chrome.csv 
Success rate: 0.67
Latency for passing runs: 21.694058240349612
Latency for failing runs: 47.30725812189507
```
 - firefox + playwright results (latency improves, accuracy stays the same)
```bash
Chetans-MacBook-Pro:browser-automation chetanmishra$ python scripts/generate_summary_stats.py --path benchmark-n-100-docker-False-playwright-True-chrome-False.csv 
Success rate: 0.7
Latency for passing runs: 14.260120006969997
Latency for failing runs: 31.83433238665263
```
 - chrome + playwright results (latency improves even more, accuracy degrades)
```bash
Chetans-MacBook-Pro:browser-automation chetanmishra$ python scripts/generate_summary_stats.py --path benchmark-n-100-docker-False-playwright-True-chrome-True.csv 
Success rate: 0.49
Latency for passing runs: 11.649881119630775
Latency for failing runs: 20.156677157271144
```
 - firefox + playwright + docker (way better success rate, latency still much better than selenium)
```bash
Chetans-MacBook-Pro:browser-automation chetanmishra$ python scripts/generate_summary_stats.py --path benchmark-n-100-docker-True-playwright-True-chrome-False.csv 
Success rate: 0.88
Latency for passing runs: 17.60762930187312
Latency for failing runs: 34.73990807930628
Chetans-MacBook-Pro:browser-automation chetanmishra$ python scripts/generate_summary_stats.py --path benchmark-n-500-docker-True-playwright-True-chrome-False.csv 
Success rate: 0.796
Latency for passing runs: 18.074777888892285
Latency for failing runs: 37.201056368210736
```
 - chrome + playwright + docker (slightly lower accuracy than firefox, but still improvement, good latency)
```bash
Chetans-MacBook-Pro:browser-automation chetanmishra$ python scripts/generate_summary_stats.py --path benchmark-n-100-docker-True-playwright-True-chrome-True.csv 
Success rate: 0.7
Latency for passing runs: 17.129104876518248
Latency for failing runs: 24.448646831512452
```
 - Frequent playwright+chrome/firefox errors:
```bash
Traceback (most recent call last):
  File "scripts/update_dob_playwright.py", line 70, in <module>
    update_dob(args.new_dob, args.browser)
  File "scripts/update_dob_playwright.py", line 42, in update_dob
    page.type('input[aria-labelledby="zf-name-label"]', "deborah")
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/sync_api/_generated.py", line 6931, in type
    noWaitAfter=no_wait_after,
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_sync_base.py", line 103, in _sync
    return task.result()
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_page.py", line 689, in type
    return await self._main_frame.type(**locals_to_params(locals()))
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_frame.py", line 474, in type
    await self._channel.send("type", locals_to_params(locals()))
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 36, in send
    return await self.inner_send(method, params, False)
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 47, in inner_send
    result = await callback.future
playwright._impl._api_types.TimeoutError: Timeout 30000ms exceeded.
```
```bash
Traceback (most recent call last):
  File "scripts/update_dob_playwright.py", line 70, in <module>
    update_dob(args.new_dob, args.browser)
  File "scripts/update_dob_playwright.py", line 39, in update_dob
    "https://secure.zenefits.com/accounts/access-trial/d1b08335-2195-4b25-bdc3-2c01ed14c017"
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/sync_api/_generated.py", line 5777, in goto
    url=url, timeout=timeout, waitUntil=wait_until, referer=referer
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_sync_base.py", line 103, in _sync
    return task.result()
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_page.py", line 464, in goto
    return await self._main_frame.goto(**locals_to_params(locals()))
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_frame.py", line 117, in goto
    await self._channel.send("goto", locals_to_params(locals()))
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 36, in send
    return await self.inner_send(method, params, False)
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 47, in inner_send
    result = await callback.future
playwright._impl._api_types.TimeoutError: Timeout 30000ms exceeded.
=========================== logs ===========================
navigating to "https://secure.zenefits.com/accounts/access-trial/d1b08335-2195-4b25-bdc3-2c01ed14c017", waiting until "load"
```
```bash
[WARNING] (benchmark_success.run_benchmarks) Run 78 failed
Traceback (most recent call last):
  File "scripts/update_dob_playwright.py", line 70, in <module>
    update_dob(args.new_dob, args.browser)
  File "scripts/update_dob_playwright.py", line 43, in update_dob
    page.click("text=Bennett, Deborah")
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/sync_api/_generated.py", line 6374, in click
    noWaitAfter=no_wait_after,
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_sync_base.py", line 103, in _sync
    return task.result()
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_page.py", line 601, in click
    return await self._main_frame.click(**locals_to_params(locals()))
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_frame.py", line 377, in click
    await self._channel.send("click", locals_to_params(locals()))
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 36, in send
    return await self.inner_send(method, params, False)
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 47, in inner_send
    result = await callback.future
playwright._impl._api_types.Error: Protocol error (DOM.scrollIntoViewIfNeeded): Cannot find context with specified id
=========================== logs ===========================
waiting for selector "text=Bennett, Deborah"
  selector resolved to visible <a rel="noreferrer noopener " href="/dashboard/#/re…>Bennett, Deborah</a>
attempting click action
  waiting for element to be visible, enabled and stable
  element is visible, enabled and stable
  scrolling into view if needed
============================================================
```
```bash
Traceback (most recent call last):
  File "scripts/update_dob_playwright.py", line 70, in <module>
    update_dob(args.new_dob, args.browser)
  File "scripts/update_dob_playwright.py", line 43, in update_dob
    page.click("text=Bennett, Deborah")
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/sync_api/_generated.py", line 6374, in click
    noWaitAfter=no_wait_after,
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_sync_base.py", line 103, in _sync
    return task.result()
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_page.py", line 601, in click
    return await self._main_frame.click(**locals_to_params(locals()))
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_frame.py", line 377, in click
    await self._channel.send("click", locals_to_params(locals()))
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 36, in send
    return await self.inner_send(method, params, False)
  File "/Users/chetanmishra/.pyenv/versions/3.7.3/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 47, in inner_send
    result = await callback.future
playwright._impl._api_types.Error: Protocol error (DOM.scrollIntoViewIfNeeded): Cannot find context with specified id
=========================== logs ===========================
waiting for selector "text=Bennett, Deborah"
  selector resolved to visible <a rel="noreferrer noopener " href="/dashboard/#/re…>Bennett, Deborah</a>
attempting click action
  waiting for element to be visible, enabled and stable
  element is visible, enabled and stable
  scrolling into view if needed
============================================================
```
```
Traceback (most recent call last):
  File "scripts/update_dob_playwright.py", line 70, in <module>
    update_dob(args.new_dob, args.browser)
  File "scripts/update_dob_playwright.py", line 43, in update_dob
    page.click("text=Bennett, Deborah")
  File "/usr/local/lib/python3.7/site-packages/playwright/sync_api/_generated.py", line 6374, in click
    noWaitAfter=no_wait_after,
  File "/usr/local/lib/python3.7/site-packages/playwright/_impl/_sync_base.py", line 103, in _sync
    return task.result()
  File "/usr/local/lib/python3.7/site-packages/playwright/_impl/_page.py", line 601, in click
    return await self._main_frame.click(**locals_to_params(locals()))
  File "/usr/local/lib/python3.7/site-packages/playwright/_impl/_frame.py", line 377, in click
    await self._channel.send("click", locals_to_params(locals()))
  File "/usr/local/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 36, in send
    return await self.inner_send(method, params, False)
  File "/usr/local/lib/python3.7/site-packages/playwright/_impl/_connection.py", line 47, in inner_send
    result = await callback.future
playwright._impl._api_types.Error: Execution context was destroyed, most likely because of a navigation.
=========================== logs ===========================
waiting for selector "text=Bennett, Deborah"
  selector resolved to visible <a rel="noreferrer noopener " href="/dashboard/#/re…>Bennett, Deborah</a>
attempting click action
  waiting for element to be visible, enabled and stable
```

### CODE REVIEW

File: prototypes/browser-automation/scripts/update_dob_playwright.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        page.goto(
+            "https://secure.zenefits.com/accounts/access-trial/d1b08335-2195-4b25-bdc3-2c01ed14c017"
+        )
+        page.goto("https://www.zenefits.com/dashboard/#/employeedirectory")
+        page.type('input[aria-labelledby="zf-name-label"]', "deborah")
+        page.click("text=Bennett, Deborah")
+        page.click('button[aria-label="Edit"]')
+        page.fill("id=dob", new_dob)
+        page.press("id=dob", "Enter")
+        page.wait_for_selector("id=dob", state="detached")
```
Comment (Oceanliu) [inline prototypes/browser-automation/scripts/update_dob_playwright.py:50]: the interface is a lot more intuitive 👍

File: prototypes/browser-automation/scripts/update_dob_playwright.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    def __str__(self):
+        return self.value
+
+
+def update_dob(new_dob, browser):
+    cookies = [
+        {
+            "name": "democenter",
+            "value": '{"accessToken":"d1b08335-2195-4b25-bdc3-2c01ed14c017","companySize":3,"startTimestamp":"2021-04-22T01:54:54.182Z","endTimestamp":"2021-05-06T01:54:54.183Z","email":"chetan@instabase.com","company_id":474596,"user_id":11375537,"employee_id":20233673,"name":"Chetan Mishra","companyName":"Instabase","firstName":"Chetan","lastName":"Mishra","companyZip":95343}',
```
Comment (Oceanliu) [inline prototypes/browser-automation/scripts/update_dob_playwright.py:20]: I can understand the access token, but all all the personal info needed for cookie?

File: prototypes/browser-automation/Makefile [hunk 3]
```diff
@@ -0,0 +1,13 @@
+BACKEND_IMAGE_NAME=backend-browtomation
```
Comment (naveenkashyap) [inline prototypes/browser-automation/Makefile:1]: `backend-brotomation` hahaha this has to be the name of the service we create if we end up doing something like this.

**Review by naveenkashyap — APPROVED** [review]

File: prototypes/browser-automation/scripts/update_dob_playwright.py (hunk 2 above)
Comment (cmishra) [reply prototypes/browser-automation/scripts/update_dob_playwright.py:20]: This won't actually be needed with a non-trial account, in that case we would just need to store user credentials (obviously not in github).

File: prototypes/browser-automation/scripts/update_dob_playwright.py (hunk 1 above)
Comment (cmishra) [reply prototypes/browser-automation/scripts/update_dob_playwright.py:50]: yup ❤️

File: prototypes/browser-automation/Makefile (hunk 3 above)
Comment (cmishra) [reply prototypes/browser-automation/Makefile:1]: +1

**Pull request merged**